helix-core/src/main/java/org/apache/helix/common/caches/CurrentStateSnapshot.java

```diff
@@ -32,18 +37,32 @@ public CurrentStateSnapshot(final Map<PropertyKey, CurrentState> currentStateMap
     if (_updatedStateKeys != null && _prevStateMap != null) {
       // Note if the prev state map is empty, this is the first time refresh.
       // So the update is not considered as "recent" change.
+      int driftCnt = 0; // clock drift count for comparing timestamp
       for (PropertyKey propertyKey : _updatedStateKeys) {
         CurrentState prevState = _prevStateMap.get(propertyKey);
         CurrentState curState = _properties.get(propertyKey);
 
         Map<String, Long> partitionUpdateEndTimes = null;
         for (String partition : curState.getPartitionStateMap().keySet()) {
           long newEndTime = curState.getEndTime(partition);
-          if (prevState == null || prevState.getEndTime(partition) < newEndTime) {
+          if (prevState == null
+              || prevState.getEndTime(partition) < newEndTime && prevState.getEndTime(partition) != -1) {
```

Comment:
Why prevState.getEndTime(partition) != -1 ? It seems that we are not using the previous end time to do any math anyway. In this case, the newly updated end time is still valid?

Reply:
The thing is that if preState did not have endTime. The getEndTime will return -1. Later when this is used to record propagation latency. The would generate exceedingly large time delta, say 1.4T as we saw in the current database test. This is to fix this bug.